cmd/goal/application.go

```diff
@@ -928,17 +929,146 @@ func printable(str string) bool {
 	return true
 }
 
+func jsonPrintable(str string) bool {
+	// htmlSafeSet holds the value true if the ASCII character with the given
+	// array position can be safely represented inside a JSON string, embedded
+	// inside of HTML <script> tags, without any additional escaping.
+	//
+	// All values are true except for the ASCII control characters (0-31), the
+	// double quote ("), the backslash character ("\"), HTML opening and closing
+	// tags ("<" and ">"), and the ampersand ("&").
+	var htmlSafeSet = [utf8.RuneSelf]bool{
+		' ':      true,
+		'!':      true,
+		'"':      false,
+		'#':      true,
+		'$':      true,
+		'%':      true,
+		'&':      false,
+		'\'':     true,
+		'(':      true,
+		')':      true,
+		'*':      true,
+		'+':      true,
+		',':      true,
+		'-':      true,
+		'.':      true,
+		'/':      true,
+		'0':      true,
+		'1':      true,
+		'2':      true,
+		'3':      true,
+		'4':      true,
+		'5':      true,
+		'6':      true,
+		'7':      true,
+		'8':      true,
+		'9':      true,
+		':':      true,
+		';':      true,
+		'<':      false,
+		'=':      true,
+		'>':      false,
+		'?':      true,
+		'@':      true,
+		'A':      true,
+		'B':      true,
+		'C':      true,
+		'D':      true,
+		'E':      true,
+		'F':      true,
+		'G':      true,
+		'H':      true,
+		'I':      true,
+		'J':      true,
+		'K':      true,
+		'L':      true,
+		'M':      true,
+		'N':      true,
+		'O':      true,
+		'P':      true,
+		'Q':      true,
+		'R':      true,
+		'S':      true,
+		'T':      true,
+		'U':      true,
+		'V':      true,
+		'W':      true,
+		'X':      true,
+		'Y':      true,
+		'Z':      true,
+		'[':      true,
+		'\\':     false,
+		']':      true,
+		'^':      true,
+		'_':      true,
+		'`':      true,
+		'a':      true,
+		'b':      true,
+		'c':      true,
+		'd':      true,
+		'e':      true,
+		'f':      true,
+		'g':      true,
+		'h':      true,
+		'i':      true,
+		'j':      true,
+		'k':      true,
+		'l':      true,
+		'm':      true,
+		'n':      true,
+		'o':      true,
+		'p':      true,
+		'q':      true,
+		'r':      true,
+		's':      true,
+		't':      true,
+		'u':      true,
+		'v':      true,
+		'w':      true,
+		'x':      true,
+		'y':      true,
+		'z':      true,
+		'{':      true,
+		'|':      true,
+		'}':      true,
+		'~':      true,
+		'\u007f': true,
+	}
+
+	for _, r := range str {
+		if r >= utf8.RuneSelf {
+			return false
+		}
+		if htmlSafeSet[r] == false {
+			return false
+		}
+	}
+	return true
+}
+
 func heuristicFormatStr(str string) string {
-	if printable(str) {
+	// if it's a 32 byte string which has a valid checksum, then make the assumption it's an address
+	if len(str) == 32 {
```

Comment:
Do you mean 58? `len('XFVTOHVXBJRMGMDNUGC7KA3IFK7QCJ2F7FMZDZOEC4TIUKEXIE6FID6S3U')==58`

Reply:
no; if you'll look where the data is coming from, you'll see that it's the byte slice stored in the TEALValue, which contains an address. The use case which you refer to is the one below, where this is printable.

Reply:
This len(str)==32 clause and the one on line 1065 do opposite things, and I think only one of them can be right.
This one parses from base32 text to a [32]byte Address, the other one encodes to base32. Lines 1065-1069 assume that a [32]byte or []byte has been typecast to string and that if we typecast it back and copy out the bytes we get an Address.
Here I don't expect 32 chars of base32 encoded data will come to anything meaningful, but 58 chars is the base32 encoded length of a [32]byte Address with checksum.
If this code works, I expect the Unmarshal() hits the `err != nil` case, falls through, and eventually just returns exactly the text that was passed in (which may happen to have been a base32 Address-with-checksum).

Reply:
I believe you're correct. Let me revise it.

Reply:
updated. thanks for the feedback!